torch/csrc/jit/tensorexpr/kernel.cpp

```diff
@@ -1482,17 +1477,127 @@ BlockPtr TensorExprKernel::bindAllInputs() {
   return block;
 }
 
-void TensorExprKernel::compile() {
-  GRAPH_DUMP("TensorExprKernel graph:", graph_);
+void TensorExprKernel::deduceMemoryLayoutPolicy() {
+  // If the tensor is channels-last contiguous, the preferred memory layout
+  // propagation policy is to use channes-last. Otherwise, the preferred policy
+  // is to use contiguous.
+  auto _prefer_symbolic_mem =
+      [](const torch::jit::Value* val,
+         const std::vector<torch::jit::StrideInput>& stride_desc_vec) {
+        TORCH_INTERNAL_ASSERT(stride_desc_vec.size() > 0);
+        // Has symbolic stride information
+        auto cur_stride_desc = stride_desc_vec[0];
+        return (cur_stride_desc ==
+                torch::jit::StrideInput::TENSOR_CONT_CHANNELS_LAST)
+            ? MemoryLayoutPolicy::kChannelsLastNdContiguous
+            : MemoryLayoutPolicy::kContiguous;
+      };
+
+  auto _prefer_static_mem = [](const torch::jit::Value* val) {
+    // No shape info is present in the graph
+    TORCH_INTERNAL_ASSERT(
+        val->isCompleteTensor(),
+        buildErrorMessage(val->debugName() + " is not a complete tensor."));
+    const auto& tt = val->type()->expect<TensorType>();
+    const auto sizes = *tt->sizes().concrete_sizes();
+    const auto strides = *tt->strides().concrete_sizes();
+    return (c10::is_channels_last_strides_2d(sizes, strides))
+        ? MemoryLayoutPolicy::kChannelsLastNdContiguous
```

Comment:
We're checking for strides_2d, but then set mem layour policy to Nd. Is it intentional? Will it work for non-2d cases?

Reply:
I'm going to support 1d and 3d. The draft code could be as follows.
```c++
    return (c10::is_channels_last_strides_2d(sizes, strides)) || (c10::is_channels_last_strides_3d(sizes, strides))
        ? MemoryLayoutPolicy::kChannelsLastNdContiguous
```